Signup tweaks

lucaslcode opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
closes #1625 

- [x] Better titles
- [x] No pre-scrolled forms
- [x] Feedback form fixes
- [x] Community guidelines

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
**Backend checklist**
- [x] Formatted my code by running `autoflake -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
  - I don't have autoflake
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [x] Added migrations if there are any database changes, ~~rebased onto `develop` if necessary for linear migration history~~

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

lucaslcode committed:
start signup tweaks

lucaslcode commented:
This can be reviewed ~~except for Jail-related files~~

lucaslcode commented on app/frontend/src/components/ContributorForm/ContributorForm.tsx:
```diff
@@ -218,60 +199,94 @@ export default function ContributorForm({ processForm }: ContributorFormProps) {
-            htmlFor="expertise"
-            component="label"
-            className={classes.label}
+          <Collapse
```
This makes the background image zoom in when the page body gets larger because it is set to `background-size: cover`. Any ideas?

darrenvong replied:
Hmm sound strange... is this CSS rule set by MUI somewhere?

lucaslcode replied:
No its just for the background image to cover the whole area

lucaslcode replied:
Fixed it by making background fixed

darrenvong commented:
Build failed - there's some import error. Can this be reviewed through the preview site (assuming the new sign up backend changes has been merged)?

darrenvong reviewed (commented):
Code looks fine to me so far, although can't see the app due to compile error so don't fully know what's up with the background zoom thing from just looking at the code :p

lucaslcode committed:
complete Community Guidelines jail section

lucaslcode committed:
implement community guidelines jail in backend

lucaslcode committed:
migrations and typos

lucaslcode committed:
cleanup

ShmuelTreiger reviewed (commented):
Can't unaccept -->must accept? Can't reject? Can't refuse?

Out of curiosity, what's the difference between TOS and community guidelines?

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -181,6 +181,7 @@ class User(Base):
     camping_ok = Column(Boolean, nullable=True)
 
     accepted_tos = Column(Integer, nullable=False, default=0)
+    accepted_community_guidelines = Column(Boolean, nullable=False, server_default="false")
```
Could we also make the CG an integer, for same reasons as TOS? Otherwise backend looks fine to me.

lucaslcode replied:
I thought it was pointless because unlike TOS, we aren't getting it from the backend

kalvdans replied:
We need to handle versions of guidelines so make it an int.

lucaslcode committed:
Better scrolling
Make radios and checkboxes primary by default
Autofocus first fields of signup
Fullwidth buttons
Use select instead of autocomplete for host status

kalvdans commented on app/proto/jail.proto:
```diff
@@ -26,20 +26,25 @@ service Jail {
     // If you're not jailed, do this with api.UpdateProfile!
   }
+
+  rpc AcceptCommunityGuidelines(AcceptCommunityGuidelinesReq) returns (JailInfoRes) {
```
Please document how I, as a third-party app developer, is going to use this method, for example by giving a link to the text to be accepted and how new versions are announced.

darrenvong commented on app/frontend/src/features/auth/signup/Signup.tsx:
```diff
@@ -181,6 +181,16 @@ export default function Signup() {
       </Hidden>
       {/***** DESKTOP ******/}
       <Hidden smDown>
+        <div className={authClasses.formCenter}>
+          <div className={authClasses.formWrapper}>
+            {error && (
+              <Alert className={authClasses.errorMessage} severity="error">
+                {error}
+              </Alert>
+            )}
+            {loading ? <CircularProgress /> : <CurrentForm />}
+          </div>
+        </div>
```
How come the markup for the form needs moving up here? Seems a bit hacky to need z-index to raise this up the page when it would have appeared at the right order where it was before based on the element order.

Also this kinda screws the natural left-right, top-down tabbing order when using a keyboard since the "Log in" link is only reachable after going through the whole form. Might be a minor accessibility concern for someone who made a mistake and got here but realise they just want to log in, but now still have to go through the whole form before they can reach the log in link.

lucaslcode replied:
I was experimenting, I'll put it back underneath

darrenvong commented on app/frontend/src/features/auth/jail/constants.ts:
```diff
@@ -1,6 +1,27 @@
+import { ReactComponent as shield } from "./resources/shield.svg";
+
+export const COMMUNITY_GUIDELINES = [
+  "Couchers.org is an inclusive community where people of all cultures and backgrounds come together. Treat everyone with mutual respect and dignity",
```
```suggestion
  "Couchers.org is an inclusive community where people of all cultures and backgrounds come together. Treat everyone with mutual respect and dignity.",
```
Missed full stop?

darrenvong commented on app/frontend/src/features/auth/jail/CommunityGuidelinesSection.tsx:
```diff
@@ -0,0 +1,149 @@
+
+        <div className={classes.grid}>
+          {COMMUNITY_GUIDELINES.map((_, index) => (
+            <React.Fragment key={`guideline-${index}`}>
```
Minor nitpick: I don't see really how adding a string prefix before it makes a difference than just use `index` alone, since it's still going to be the value of `index` that makes the keys unique, so may as well use it directly?

darrenvong commented on app/frontend/src/features/auth/jail/CommunityGuidelinesSection.tsx:
```diff
@@ -0,0 +1,149 @@
+          disabled={completed || !formState.isValid}
+          className={classes.button}
+        >
+          {completed ? "Thanks!" : "Continue"}
```
Use existing values from constants file or pull this to constants if they don't already exist?

darrenvong commented on app/frontend/src/components/ContributorForm/ContributorForm.tsx:
```diff
@@ -218,61 +209,100 @@ export default function ContributorForm({ processForm }: ContributorFormProps) {
-            className={classes.label}
+          <Collapse
+            in={
+              watchContribute !== CONTRIBUTE_OPTIONS[2] &&
```
```suggestion
              watchContribute !== "No" &&
```
I feel like in this case, this is clearer as opposed to having to check out what the third option is and an ok compromise to have the value in the file

lucaslcode replied:
I don't think that will work once we have translations. I'd rather leave this. (I did `CONTRIBUTE_OPTIONS as const` so you can just hover to see the values)

darrenvong replied:
We'll probably have to change this again anyway when we do translations, but if hovering gives the value in VS Code then let's leave it 👍

darrenvong commented on app/frontend/src/components/ContributorForm/ContributorForm.tsx:
```diff
@@ -218,61 +209,100 @@ export default function ContributorForm({ processForm }: ContributorFormProps) {
-            name="expertise"
-            helperText={errors?.expertise?.message ?? EXPERTISE_HELPER}
-            error={!!errors?.expertise?.message}
+            <Controller
+              id="contributeWays"
+              control={control}
```
I'm noticing some weird focus state styling, where the section heading changes colour when any of the checkboxes get focused:
![Screenshot 2021-07-17 at 02 18 46](https://user-images.githubusercontent.com/13669362/126021248-f0dd489b-cc80-4266-8891-89f03be99757.png)

I don't think the title should change colour to be consistent with other input headings/labels. Instead, the focused checkbox option should show up/change etc depending on which checkbox is being focused on, which isn't the case at the moment. You have to blindly tab (not knowing where you are in the list of checkboxes) until the focus goes to the "what expertise do you have..." input field.

darrenvong commented on app/frontend/src/components/ContributorForm/ContributorForm.tsx:
```diff
@@ -218,61 +209,100 @@ export default function ContributorForm({ processForm }: ContributorFormProps) {
             className={classes.textbox}
           />
           <Controller
-            id="contributeWays"
+            id="contribute"
             control={control}
```
Same thing as what I described for `contributeWays` seems to be happening here too

darrenvong commented on app/frontend/src/features/auth/jail/resources/ghost.svg:
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="100%" height="100%" clipboard-id="document_1bs6v2b"><defs>
```
These SVG icons don't look very optimised and way more complicated than needed from a glance. I think it might be worth optimising it with a tool like [SVGOMG](https://jakearchibald.github.io/svgomg/) before using them. Only tried it on this file and it's reduced the file size by almost 1/3, so helps to speed up downloading them as well!

lucaslcode committed:
signup api change for guidelines step, start implementing

lucaslcode committed:
finish implementing signup step for guidelines

lucaslcode committed:
update tests

lucaslcode committed:
add community guidelines form test

lucaslcode committed:
review comments [skip CI]

aapeliv committed:
Make guidelines version an int

aapeliv committed:
Fix models too

aapeliv committed:
Fix frontend tests and lint warnings

aapeliv committed:
Merge branch 'develop' into frontend/chore/signup-v2-tweaks

aapeliv committed:
Fix migrations

aapeliv committed:
Fix tests

aapeliv committed:
Fix up backend tests

aapeliv committed:
Implement GC for signup in backend

lucaslcode committed:
don't use workers for tests (workaround)

lucaslcode commented:
Added `--runInBand` to prevent multiple tests running at once. This is bad but I don't even know which test is causing failures

aapeliv commented:
Still failing :/

darrenvong commented on app/frontend/src/components/CommunityGuidelines.tsx/CommunityGuidelines.tsx:
```diff
@@ -0,0 +1,146 @@
+import {
```
The folder name for this component and the rest of the files shouldn't have `.tsx` in 🤔

darrenvong commented on app/frontend/src/components/ContributorForm/ContributorForm.tsx:
```diff
@@ -139,7 +151,11 @@ export default function ContributorForm({ processForm }: ContributorFormProps) {
           </Typography>
           <TextField
-            inputRef={register}
+            inputRef={(el: HTMLInputElement) => {
+              if (!ideasInputRef.current && autofocus) el.focus();
+              if (el) ideasInputRef.current = el;
+              register(el);
+            }}
```
I think it might be safer to put an if check on `el` to see if it exists first before calling `el.focus`, as [callback refs could get called with `null` first before receiving the `HTMLInputElement` instance](https://reactjs.org/docs/refs-and-the-dom.html#caveats-with-callback-refs) when using an inline function, so `el` should be really be `HTMLInputElement | null` (the MUI type here is a bit crap as it marks `el` with `any`...)

darrenvong commented on app/frontend/src/features/auth/signup/AccountForm.tsx:
```diff
@@ -170,17 +168,21 @@ export default function AccountForm() {
-            },
-          })}
+          inputRef={(el: HTMLInputElement) => {
+            if (!usernameInputRef.current) el.focus();
```
Same comment as in contributor form on `el` check

darrenvong commented on app/frontend/src/features/auth/signup/BasicForm.tsx:
```diff
@@ -65,13 +73,17 @@ export default function BasicForm() {
-            required: NAME_REQUIRED,
-          })}
+          inputRef={(el: HTMLInputElement) => {
+            if (!nameInputRef.current) el.focus();
```
Same comment as in contributor form

darrenvong commented on app/frontend/src/features/auth/signup/Signup.test.tsx:
```diff
@@ -216,10 +250,51 @@ describe("Signup", () => {
+      needBasic: false,
+      needAccount: false,
+      needAcceptCommunityGuidelines: true,
+      needFeedback: false,
```
```suggestion
      needFeedback: true,
```
Otherwise doesn't match test description?

darrenvong commented on app/frontend/src/theme.ts:
```diff
@@ -37,11 +37,9 @@ const themeOptions: ThemeOptions = {
       root: {
-        padding: 0,
-        paddingTop: spacing(1),
+        "&$focused": { color: "inherit" },
```
What does "$" do here?

lucaslcode replied:
Some MUI magic. This was to fix the colour change on focus. I tried `MuiFormLabel: { focused: { color: "inherit" } }` and I got a helpful warning telling me to do what is here now

lucaslcode committed:
review and fixes

aapeliv committed:
Try -w 1

aapeliv committed:
Remove --runInBand
